core/lib/include/xfa.h

```diff
@@ -42,16 +42,28 @@ _Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
  *
  * @internal
  */
-#define _XFA(name, prio) __attribute__((used, section(".xfa." #name "." #prio)))
+#if defined(__llvm__) || defined(__clang__)
+#define _XFA(name, prio) \
+    __attribute__((no_sanitize("address"))) \
+    __attribute__((used, section(".xfa." #name "." #prio)))
+#else
+#define _XFA(name, prio) \
+    __attribute__((used, section(".xfa." #name "." #prio)))
+#endif
```

Comment:
```suggestion
#if defined(__llvm__) || defined(__clang__)
#define NO_SANITIZE __attribute__((no_sanitize("address")))
#endif

#define _XFA(name, prio) \
    NO_SANITIZE \
    __attribute__((used, section(".xfa." #name "." #prio)))
```

would be a bit more readable

Reply:
I thought about doing it that way myself but I spotted an issue with this approach. The `NO_SANITIZE` compiler hint looks like it disables ASAN - which it only does when using llvm. That would be misleading for a gnu user. One could try to alter the hint in a way that also disables ASAN for gnu but the documentation states that the gnu compiler attribute only applies to functions. This is in contrast to clang, where it also applies to global variables.

see [gnu docs](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html)
see [clang docs](https://clang.llvm.org/docs/AttributeReference.html#no-sanitize)

Reply:
But is it also *needed* here with GCC?
If adding the `no_sanitize` attribute here also for GCC, I don't see why we should special-case Clang.

Reply:
I think I might misunderstood your intentions? Just removing the gcc/clang special-case results in gcc complaining ` error: ‘no_sanitize_address’ attribute ignored [-Werror=attributes]`

Reply:
Ah I see.
So maybe then `NO_SANITIZE_ARRAY` that is simply a no-op on GCC (I assume it will never sanitize array access then).
If it gains the ability to sanitize things other than functions, we can also add an implementation for GCC and leave it a no-op until then.

Reply:
Like this?